presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+
+        if (bytes.length > UNSCALED_DECIMAL_128_SLICE_LENGTH
+                || (bytes.length == UNSCALED_DECIMAL_128_SLICE_LENGTH && (bytes[0] & SIGN_BYTE_MASK) != 0)) {
+            throwOverflowException();
+        }
+
+        // convert to little-endian order
+        reverse(bytes);
+
+        Slice decimal = Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        decimal.setBytes(0, bytes);
+        if (unscaledValue.signum() < 0) {
+            setNegative(decimal, true);
+        }
+
+        throwIfOverflows(decimal);
+
+        return decimal;
+    }
+
+    public static Slice unscaledDecimal(long unscaledValue)
+    {
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
+        boolean negative = isNegative(decimal);
+
+        if (hi != 0 || ((lo > Long.MIN_VALUE || !negative) && lo < 0)) {
+            throwOverflowException();
+        }
+
+        return negative ? -lo : lo;
+    }
+
+    public static long unscaledDecimalToUnscaledLongUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        return isNegative(decimal) ? -lo : lo;
+    }
+
+    public static Slice rescale(Slice decimal, int rescaleFactor)
+    {
+        if (rescaleFactor == 0) {
+            return decimal;
+        }
+        else {
+            Slice result = unscaledDecimal();
+            rescale(decimal, rescaleFactor, result);
+            return result;
+        }
+    }
+
+    public static void rescale(Slice decimal, int rescaleFactor, Slice result)
+    {
+        if (rescaleFactor == 0) {
+            copyUnscaledDecimal(decimal, result);
+        }
+        else if (rescaleFactor > 0) {
+            if (rescaleFactor >= POWERS_OF_TEN.length) {
+                throwOverflowException();
+            }
+            multiply(decimal, POWERS_OF_TEN[rescaleFactor], result);
+        }
+        else {
+            scaleDownRoundUp(decimal, -rescaleFactor, result);
+        }
+    }
+
+    public static Slice multiply(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+        multiply(left, right, result);
+        return result;
+    }
+
+    public static void multiply(Slice left, Slice right, Slice result)
+    {
+        int l0 = getInt(left, 0);
```

Comment:
These are the results from a quick benchmark of the multiply method after applying these changes and the ones below:

```
before: 66735.429 ± 2042.083  ops/ms
after:  73274.755 ± 2015.445  ops/ms
```